### PR TITLE
Store event data in a map

### PR DESCRIPTION
**Pull request by yanivagman**
## Description (git log)

    Currently event data is stored on the stack using a struct named
    event_data_t. This has the drawback of consuming 120 bytes
    (sizeof(event_data_t)) out of the available 512 bytes before we even
    begin considering the required logic of the event. By moving this struct
    into a map, not only we save this space on the stack, but we can then
    also add a buffer to save the event's arguments instead of having a
    dedicated submit buffer. This allows us to place the buffer right after
    the context of the event, and send it to userspace without having to do
    an extra copy of the context to the submit buffer.
    
    In addition to the above benefits, this also sets the ground to split
    our bpf programs into several programs linked by tail calls and keeping
    the event's state on the event data map.

## Type of change

- [x] Code refactor (code improvement and/or code removal)

## How Has This Been Tested?

Tests being included in this PR:

- [ ] Test File A
- [ ] Test File B

Reproduce the test by running:

- command 01
- command 02

## Final Checklist:

Pick "Bug Fix" or "Feature", delete the other and mark appropriate checks.

- [ ] I have made corresponding changes to the documentation.
- [ ] My code follows the style guidelines (C and Go) of this project.
- [ ] I have performed a self-review of my own code.
- [ ] I have commented all functions/methods created explaining what they do.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] My changes generate no new warnings.
- [ ] I have added tests that prove my fix, or feature, is effective.
- [ ] New and existing unit tests pass locally with my changes.
- [ ] Any dependent changes have been merged and published before.

## Git Log Checklist:

My commits logs have:

- [ ] Subject starts with "subsystem|file: description".
- [ ] Do not end the subject line with a period.
- [ ] Limit the subject line to 50 characters.
- [ ] Separate subject from body with a blank line.
- [ ] Use the imperative mood in the subject line.
- [ ] Wrap the body at 72 characters.
- [ ] Use the body to explain what and why instead of how.

### CODE REVIEW

**Review by NDStrahilevitz — COMMENTED** [review]
I want to run this through the eBPF benchmark test we have to see if there's a regression or improvement here.

**Comment by rafaeldtinoco** [comment]
> I want to run this through the eBPF benchmark test we have to see if there's a regression or improvement here.

And I want to check the network code before +1 (just to make sure), and give a second fast look after Nadav.

**Comment by yanivagman** [comment]
> Did some benchmarking for this, overall looks like this will be a rather significant degradation.

I run the reddis-benchmark locally and I don't see any difference between main and this branch. Even slightly better results with this branch. Can you try to running the tests in the opposite order? It might be related to some load you had or caching

File: pkg/ebpf/c/tracee.bpf.c
Comment (yanivagman) [reply pkg/ebpf/c/tracee.bpf.c:1968]: Nope. We only init event_data in the submit phase (to optimize the init phase), so there's no place to use it there

File: pkg/ebpf/c/tracee.bpf.c
Comment (yanivagman) [reply pkg/ebpf/c/tracee.bpf.c:858]: Choosing names is the most time consuming part in writing PRs.
The idea of this struct is to hold pointers to the maps we use in our programs, and potentialy other per-program specific state (opposed to per-event state, for which we use the event_data_t). I'm not sure the config map can be considered as "trace data", yet I open for other suggestion to rename this struct.

Comment (yanivagman) [reply pkg/ebpf/c/tracee.bpf.c]: Renamed to program_data_t

**Comment by NDStrahilevitz** [comment]
> I run the reddis-benchmark locally and I don't see any difference between main and this branch. Even slightly better results with this branch. Can you try to running the tests in the opposite order? It might be related to some load you had or caching

Ran benchmark again on a rebooted machine, still getting slightly worse results with the branch but it might be in error range. I think we will need a dedicated machine to know for sure at this point, or `redis-benchmark` might not be such a reliable tool for us after all.
Anyway i'm ok with merging all things considered..

**Review by NDStrahilevitz — APPROVED** [review]

**Comment by yanivagman** [comment]
> > I want to run this through the eBPF benchmark test we have to see if there's a regression or improvement here.
> 
> And I want to check the network code before +1 (just to make sure), and give a second fast look after Nadav.

@rafaeldtinoco waiting for your approval before I merge

**Review by rafaeldtinoco — APPROVED** [review]
LGTM @yanivagman

**Pull request merged**
